src/herder/test/HerderTests.cpp

```diff
@@ -274,138 +287,140 @@ testTxSet(uint32 protocolVersion)
     int64_t amountPop =
         nbTransactions * app->getLedgerManager().getLastTxFee() + minBalance0;
 
-    TxSetFramePtr txSet = std::make_shared<TxSetFrame>(Hash{});
+    TxSetFrameConstPtr txSet = std::make_shared<TxSetFrame const>(Hash{});
+    std::vector<TransactionFrameBasePtr>
+        txOriginal; // Keep the order of the generated transactions
 
     auto genTx = [&](int nbTxs) {
         std::string accountName = fmt::format("A{}", accounts.size());
         accounts.push_back(root.create(accountName.c_str(), amountPop));
         auto& account = accounts.back();
+        std::vector<TransactionFrameBasePtr> txs;
         for (int j = 0; j < nbTxs; j++)
         {
             // payment to self
-            txSet->add(account.tx({payment(account.getPublicKey(), 10000)}));
+            txs.push_back(account.tx({payment(account.getPublicKey(), 10000)}));
         }
+        txSet = TxSetUtils::addTxs(txSet, txs);
+        txOriginal.insert(txOriginal.end(), txs.begin(), txs.end());
     };
-
     for (size_t i = 0; i < nbAccounts; i++)
     {
         genTx(nbTransactions);
     }
-
-    txSet->previousLedgerHash() =
-        app->getLedgerManager().getLastClosedLedgerHeader().hash;
+    txSet = std::make_shared<TxSetFrame const>(
+        app->getLedgerManager().getLastClosedLedgerHeader().hash,
+        txSet->getTxsInHashOrder());
 
     SECTION("too many txs")
     {
-        while (txSet->mTransactions.size() <=
-               cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE)
+        while (txSet->sizeTx() <= cfg.TESTING_UPGRADE_MAX_TX_SET_SIZE)
         {
             genTx(1);
         }
-        txSet->previousLedgerHash() =
-            app->getLedgerManager().getLastClosedLedgerHeader().hash;
-        txSet->sortForHash();
+        txSet = std::make_shared<TxSetFrame const>(
+            app->getLedgerManager().getLastClosedLedgerHeader().hash,
+            txSet->getTxsInHashOrder());
         REQUIRE(!txSet->checkValid(*app, 0, 0));
     }
     SECTION("order check")
     {
-        txSet->sortForHash();
-
         SECTION("success")
         {
             REQUIRE(txSet->checkValid(*app, 0, 0));
 
-            txSet->trimInvalid(*app, 0, 0);
+            trimInvalid(*app, txSet, 0, 0);
             REQUIRE(txSet->checkValid(*app, 0, 0));
-            REQUIRE(txSet->mTransactions.size() == (2 * nbTransactions));
+            REQUIRE(txSet->sizeTx() == (2 * nbTransactions));
         }
+        /* No longer possible for txs to be out of order, and transactions are
+        no longer mutable.
```

Comment:
I'll remove this block once I get a sign-off.

Reply:
I've re-enabled this test to cover the case of an out-of-order `TransactionSet`.